src/vehicle.cpp

```diff
@@ -6317,36 +6281,98 @@ std::set<tripoint> &vehicle::get_points( const bool force_refresh )
+    const int py = modulo( real_global_pos.y, SEEY );
+    auto sm = MAPBUFFER.lookup_submap( smp );
+    if( sm == nullptr ) {
+        debugmsg( "is_sm_tile_outside(): couldn't find submap %d,%d,%d", smp.x, smp.y, smp.z );
+        return false;
+    }
+
+    if( px < 0 || px >= SEEX || py < 0 || py >= SEEY ) {
+        debugmsg("err %d,%d", px, py);
+        return false;
+    }
+
+    return !(sm->ter[px][py].obj().has_flag(TFLAG_INDOORS) ||
+        sm->get_furn(px, py).obj().has_flag(TFLAG_INDOORS));
+}
+
+void vehicle::update_time( const calendar &update_to )
 {
-    tripoint veh_loc = global_pos3();
-    // Don't fill funnels every turn, because rainfall has 10 turn granularity
-    if( smz >= 0 && !funnels.empty() && ( calendar::turn - last_update_turn >= 10 || one_in( 10 ) ) ) {
+    if( smz < 0 ) {
+        return;
+    }
+
+    // Weather stuff, only for z-levels >= 0
+    // TODO: Have it wash cars from blood?
+    if( funnels.empty() && solar_panels.empty() ) {
+        return;
+    }
+
+    const auto update_from = last_update_turn;
+    if( update_to - update_from < MINUTES(1) ) {
+        // We don't need to check every turn
+        return;
+    }
+
+    last_update_turn = update_to;
+
+    // Get one weather data set per veh, they don't differ much across veh area
+    const tripoint veh_loc = real_global_pos3();
+    auto accum_weather = sum_conditions( update_from, update_to, veh_loc );
+    if( !funnels.empty() ) {
         double rain_amount = 0.0;
         // TODO: double acid_amount = 0.0;
-        for( int fun : funnels ) {
-            tripoint location = veh_loc + parts[fun].precalc[0];
-            // Can't use g->is_sheltered
-            // TODO: Fix procing vehicles partially out of map
-            if( g->m.has_flag( TFLAG_INDOORS, location ) ) {
+        for( int part : funnels ) {
+            const tripoint part_loc = veh_loc + parts[part].precalc[0];
+            if( !is_sm_tile_outside( part_loc ) ) {
                 continue;
             }
 
-            rainfall_data rainfall = get_rainfall( last_update_turn, calendar::turn, location );
-
-            const int part_size = part_info( fun ).size;
+            const int part_size = part_info( part ).size;
             const double funnel_area_mm = M_PI * part_size * part_size;
-
-            rain_amount += funnel_charges_per_turn( funnel_area_mm, rainfall.rain_amount );
+            rain_amount += funnel_charges_per_turn( funnel_area_mm, accum_weather.rain_amount );
         }
 
         const int rain_val = divide_roll_remainder( rain_amount, 1.0 );
         if( rain_val > 0 ) {
             refill( "water", rain_val );
+            add_msg( m_debug, "%s got %d water from funnels", name.c_str(), rain_val );
         }
     }
 
-    last_update_turn = calendar::turn;
+    if( !solar_panels.empty() ) {
+        int epower = 0;
+        for( int part : solar_panels ) {
+            if( parts[part].hp <= 0 ) {
```

Comment:
Strange asymmetry: the funnel loop has no such check for its HP, broken funnels will just work.